src/client/datascience/jupyter/kernels/cellDisplayIdTracker.ts

```diff
@@ -0,0 +1,63 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+import { injectable, inject } from 'inversify';
+import { NotebookCell, NotebookCellOutput, NotebookDocument } from 'vscode';
+import { IVSCodeNotebook } from '../../../common/application/types';
+import { createDeferred, Deferred } from '../../../common/utils/async';
+
+@injectable()
+export class CellOutputDisplayIdTracker {
+    private displayIdCellOutputMappingPerDocument = new WeakMap<
+        NotebookDocument,
+        Map<string, { output: Deferred<NotebookCellOutput>; cell: NotebookCell }>
+    >();
+    private cellToDisplayIdMapping = new WeakMap<NotebookCell, string>();
+    constructor(@inject(IVSCodeNotebook) notebooks: IVSCodeNotebook) {
+        notebooks.onDidChangeNotebookDocument((e) => {
+            // We are only interested in cells that were cleared
+            if (e.type === 'changeCellOutputs') {
+                e.cells
+                    .filter((cell) => cell.outputs.length)
+                    .map((cell) => {
+                        // If a cell was cleared, then remove the mapping, the output cannot exist anymore.
+                        const displayIdToDelete = this.cellToDisplayIdMapping.get(cell);
+                        if (displayIdToDelete) {
+                            this.cellToDisplayIdMapping.delete(cell);
+                            this.displayIdCellOutputMappingPerDocument.get(cell.notebook)?.delete(displayIdToDelete);
+                        }
+                    });
+            }
+        });
+    }
+    /**
+     * Keep track of the mapping between display_id and the output.
+     * When we need to update this display, we can resolve the promise & access the output.
+     * The return value is a promise that needs to be resolved with the associated output thats been added to the DOM
+     */
+    public trackOutputByDisplayId(cell: NotebookCell, displayId: string): Deferred<NotebookCellOutput> {
```

Comment:
I might just need to read this over a second time, but I can't quite tell if it's mapping to the Cell or the Output item here.

Reply:
Both?

Map from document to display id map of cells and a promise
Map from cell to display id

Reply:
Yeah, I think I got it on the second pass. Wasn't really parsing that it was creating the deferred here in this function.

Reply:
Happy to refactor to something else

Reply:
Nah it's fine just took a bit to parse.